app/helpers/admin_helper.rb

```diff
@@ -45,4 +45,66 @@ def admin_menu_items
 
     MENU_ITEMS.dup.insert(7, PROFILE_ADMIN)
   end
+
+  # @ridhwana WIP on new admin navigation elements
+  # The outer keys in NESTED_MENU_ITEMS need to correspond to the scope of the route
+  # On second level navigation with more children, we reference the default tabs controller. i.e look at developer_tools
+  NESTED_MENU_ITEMS = {
```

Comment:
So, when I saw this I wasn't so happy with a few things, e.g. having to specify `controller` even when it's the same as the `name`. So I came up with a little abstraction for defining menus:

```ruby
class Menu
  def self.define(&block)
    builder = new
    builder.instance_exec(&block)
    builder.items
  end

  attr_reader :items

  def initialize
    @items = {}
  end

  def scope(name, children)
    @items[name] = children
  end

  def item(name:, controller: name, children: [])
    { name: name, controller: controller, children: children }
  end
end
```

This allows us to specify menus more declaratively:

```ruby
AdminMenu = Menu.define do
  scope :overview, [
    item(name: "overview")
  ]

  scope :content_manager, [
    item(name: "posts", controller: "articles"),
    item(name: "badges", children: [
      item(name: "badge library", controller: "badges")
    ])
  ]
end
```

The resulting data structure has the same format as the one you wrote here manually:

```ruby
pp AdminMenu
{:overview=>[{:name=>"overview", :controller=>"overview", :children=>[]}],
 :content_manager=>
  [{:name=>"posts", :controller=>"articles", :children=>[]},
   {:name=>"badges",
    :controller=>"badges",
    :children=>
     [{:name=>"badge library", :controller=>"badges", :children=>[]}]}]}
```

This would allow us to make this menu a  first-class citizen of the app and move the helper into it:

```ruby
class AdminMenu
  ITEMS = Menu.define do
    scope :overview, [
      item(name: "overview")
    ]

    scope :content_manager, [
      item(name: "posts", controller: "articles"),
      item(name: "badges", children: [
        item(name: "badge library", controller: "badges")
      ])
    ]
  end

  def self.get_nested_menu_items(group_name, child_nav_item)
    return ITEMS[group_name.to_sym].each do |items|
      return items if items[:controller] == child_nav_item

      if items[:children] && items[:children].length > 0
        return items[:children].each do |child|
          return items if child[:controller] == child_nav_item
        end
      end
    end
  end
end
```

This way the menu is responsible for giving us information about itself:

```ruby
AdminMenu.get_nested_menu_items("content_manager", "badges")
#=> {:name=>"badges", :controller=>"badges", :children=>[{:name=>"badge library", :controller=>"badges", :children=>[]}]}
AdminMenu.get_nested_menu_items("content_manager", "badge library")
#=> [{:name=>"badge library", :controller=>"badges", :children=>[]}]
```

This might be overkill, but at least it was fun to write. 😄 The advantage would be that we could theoretically change the underlying storage or other implementation details while not having to worry that other code breaks.

Reply:
This is great @citizen428 thanks.😊 Will look into the implementation this morning!

Reply:
Don't feel like you have to @Ridhwana, I was just brainstorming in code. But feel free to re-use this if it seems useful and more important also feel free to ask any questions you might have about the implementation.

Reply:
I do prefer this method, it's more readable, understandable and less clunky.
I think I may even whip up a little blog post about things I learned from your implementation.

Reply:
@citizen428 I wasn't sure in which folder the refactored AdminMenu and Menu file belonged. I added them as helpers but I suspect this may not be the correct place.

Reply:
@Ridhwana I'd but `Menu` in `app/lib` (we could theoretically reuse it elsewhere) and `AdminMenu` in `app/models` (we have other non-DB backed models there).

Reply:
Thank you, and also thanks for this suggestion, Im pretty chuffed with the result.